### PR TITLE
scripts: tests: Expand environment.py tests

### PR DESCRIPTION
**Pull request by LukaszMrugala**
As we aim to enhance our test coverage, the `environment` module is a target-rich environment. This addition would increase this module's coverage to over 95%.

Some errors in the original module were discovered and tests regarding these cases were commented out until they are fixed. Of course, we can use the TDD approach and merge failing tests to incentivise fixes, but I leave that to be discussed.
One of such errors is described in Issue #60035

### CODE REVIEW

**Comment by hakehuang** [comment]
one question, how those test be triggerred?

**Comment by LukaszMrugala** [comment]
> one question, how those test be triggerred?

To run tests manually just from this file, `pytest $ZEPHYR_BASE/scripts/tests/twister/test_environment.py`. It requires firstly setting `ZEPHYR_BASE`, e.g. via the `zephyr-env.sh` script.

**Comment by hakehuang** [comment]
> > one question, how those test be triggerred?
> 
> To run tests manually just from this file, `pytest $ZEPHYR_BASE/scripts/tests/twister/test_environment.py`. It requires firstly setting `ZEPHYR_BASE`, e.g. via the `zephyr-env.sh` script.

is there a way that this be automatically triggerred? per any changes to twister related scripts?

**Comment by LukaszMrugala** [comment]
> > > one question, how those test be triggerred?
> > 
> > 
> > To run tests manually just from this file, `pytest $ZEPHYR_BASE/scripts/tests/twister/test_environment.py`. It requires firstly setting `ZEPHYR_BASE`, e.g. via the `zephyr-env.sh` script.
> 
> is there a way that this be automatically triggerred? per any changes to twister related scripts?

It is automatically run in PRs already - `.github/workflows/twister_tests.yml` uses `pytest $ZEPHYR_BASE/scripts/tests/twister/` to run all `test_[...]` files. If you take a look at checks in this PRs: [Twister Unit Tests (3.8, ubuntu-22.04)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5463826059/jobs/9945009393?pr=60037#logs), [Twister Unit Tests (3.9, ubuntu-22.04)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5463826059/jobs/9945009530?pr=60037#logs), [Twister Unit Tests (3.10, ubuntu-22.04)](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5463826059/jobs/9945009650?pr=60037#logs), you can see automatic runs of `test_environment.py`

**Review by hakehuang — DISMISSED** [review]

**Review by nashif — APPROVED** [review]

File: scripts/tests/twister/test_environment.py [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+
+    with mock.patch('sys.argv', ['twister'] + args):
+        parser = twisterlib.environment.add_parse_arguments()
+
+    if which_dict:
+        which_dict['path'] = {name: shutil.which(name) \
+            for name in which_dict['exist']}
+        which_mock = mock.Mock(side_effect=mock_which)
+
+    with mock.patch('os.name', os_name) \
```
Comment (gchwier) [inline scripts/tests/twister/test_environment.py:181]:
```suggestion
    with mock.patch('os.name', os_name) if os_name is not None else nullcontext(), \
         mock.patch('shutil.which', which_mock) if which_dict else nullcontext(), \
         mock.patch('twisterlib.environment.PYTEST_PLUGIN_INSTALLED', pytest_installed_var) \
            if pytest_installed_var is not None else nullcontext():
```
indentation could be improved to make mocking code more readable.
Could be fine to use [parenthesized-context-manager](https://docs.python.org/3.10/whatsnew/3.10.html#parenthesized-context-managers)
but it is not available in python 3.8

File: scripts/tests/twister/test_environment.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+
+    mock_datetime = mock.Mock(
+        now=mock.Mock(
+            return_value=mock.Mock(
+                isoformat=mock.Mock(return_value='dummy_time')
+            )
+        )
+    )
+
+    with mock.patch.object(
```
Comment (gchwier) [inline scripts/tests/twister/test_environment.py:365]:
```suggestion
    with mock.patch.object(
            twisterlib.environment.TwisterEnv,
            'check_zephyr_version',
            mock.Mock()) as mock_czv, \
         mock.patch.object(
            twisterlib.environment.TwisterEnv,
            'get_toolchain',
            mock.Mock()) as mock_gt, \
         mock.patch('twisterlib.environment.datetime', mock_datetime):
```
indentation only

**Review by gchwier — APPROVED** [review]
only suggestions to indentations

File: scripts/tests/twister/test_environment.py (hunk 1 above)
Comment (LukaszMrugala) [reply scripts/tests/twister/test_environment.py:181]: I've tried to fit within 80 chars, but that really hurts readability in context managers.
I think that keeping to breaking a single context manager 'line' only on `if` and allowing the text to overrun the 80-char line is a good compromise for now.

File: scripts/tests/twister/test_environment.py (hunk 2 above)
Comment (LukaszMrugala) [reply scripts/tests/twister/test_environment.py:365]: I've tried to keep to 1TBS, as it is pretty clear in most cases, but I see that this indentation style is very good for the context manager.

**Pull request merged**